### PR TITLE
Add handling for components, attachments, embeds, and emoji (only preliminary). 

### PR DESCRIPTION
**Pull request by clin1234**

### CODE REVIEW

**Comment by clin1234** [comment]
Status?

**Comment by guerinoni** [comment]
I can't merge it... 
- 0bf8eeb merge instead of rebase
- Omit version number, add features for channel and message objects duplicated commit
- Add preliminary support for attachments and roles duplicated commit
- Add preliminary support for attachments and roles duplicated commit
- Replace .value call with subscript operator for settings duplicated commit
- Remove Git diff crud left by a previous commit duplicated commit
- Fix formatting issue introduced by Qt Creator duplicated commit
- Run clang-format duplicated commit
- Fix up bungles of previous commits duplicated commit
- Add preliminary support for attachments and roles duplicated commit
- Add preliminary support for attachments and roles duplicated commit
- Fix build errors, and rename role files to proper case duplicated commit
- Clean code duplicated commit
- Add latest version of API duplicated commit

**Comment by clin1234** [comment]
Should be fixed...

File: src/core/Guild.cpp [hunk 1]
```diff
@@ -79,3 +99,152 @@ void Guild::setIconBase64(QByteArray &&data)
...
+{
+    return m_explilvl;
+}
+
+bool Guild::is_only_mentioned() const
+{
+    return only_mentioned;
+}
+
+<<<<<<< HEAD
```
Comment (guerinoni) [inline src/core/Guild.cpp:163]: Build fails with this

File: src/core/Channel.cpp [hunk 2]
```diff
@@ -2,31 +2,43 @@
...
+    message_count       = obj["message_count"].toString().toUShort();
+    member_count        = obj["member_count"].toString().toUShort();
 }
 
+/*
+bool key_exists(const QJsonObject &k, QString &l)
+{
+    return !k[l).isNull();
+}*/
+
```
Comment (guerinoni) [inline src/core/Channel.cpp:25]: Remove unused code

File: src/core/Component.cpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+void Button::unmarshal(const QJsonObject &&o)
+{
+    m_url = o[QStringLiteral("url")].toString();
+    if (o.contains(QStringLiteral("style")))
+        m_style = static_cast<Button_Style>(o[QStringLiteral("style")].toInt());
+    if (o.contains(QStringLiteral("emoji"))) {
+        // Emoji t;
+        // t.unmarshal(o[QStringLiteral("emoji")].toObject());
+        // m_emoji.emplace(t);
```
Comment (guerinoni) [inline src/core/Component.cpp:65]: Unused! Add a TODO or FIXME instead and a warning print

File: src/core/Component.cpp [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    }
+}
+
+/*
+template<class T> constexpr T &get(Component &c)
+{
+    int k = c.index();
+    return std::get<k>(c);
+}
+*/
```
Comment (guerinoni) [inline src/core/Component.cpp:50]: Remove unused

File: src/core/Component.cpp [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                Button but;
+                but.unmarshal(b.toObject());
+                a.buttons()->emplace_back(but);
+            }
+        }
+        m_component.emplace<Action_Row>(a);
+    } else {
+        // Button but;
+        // but.unmarshal(o);
+        // m_component.emplace<Button>(but);
```
Comment (guerinoni) [inline src/core/Component.cpp:40]: Unused! Add a TODO or FIXME instead and a warning print

File: src/core/Component.cpp [hunk 6]
```diff
@@ -0,0 +1,126 @@
+#include "Component.hpp"
+
+#include <QJsonArray>
+#include <utility>
+
+constexpr int Component::type()
```
Comment (guerinoni) [inline src/core/Component.cpp:6]: Maybe is better an enum, but at least add a FIXME

File: src/core/Emoji.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+const QString &Emoji::name() const
+{
+    return m_name;
+}
+/*
+optional<User> &Emoji::user()
+{
+    return m_user;
+}*/
```
Comment (guerinoni) [inline src/core/Emoji.cpp:18]: Unused!

File: src/core/Emoji.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        m_managed = o["managed"].toBool();
+    if (o.contains("animated"))
+        m_animated = o["animated"].toBool();
+    if (o.contains("available"))
+        m_available = o["available"].toBool();
+    if (o.contains("user")) {
+        const auto &t{ QJsonDocument(o["user"].toObject()) };
+        // User u;
+        // u.populate(t);
+        // m_user.emplace(u);
```
Comment (guerinoni) [inline src/core/Emoji.cpp:57]: Unused! Add a TODO or FIXME instead and a warning print

File: src/core/Emoji.hpp [hunk 9]
```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "Types.hpp"
+//#include "User.hpp"
```
Comment (guerinoni) [inline src/core/Emoji.hpp:4]: Unused

File: src/core/Emoji.hpp [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+using std::optional;
+
+class Emoji
+{
+    snowflake m_id;
+    QString m_name;
+    optional<bool> m_require_colons, m_managed, m_animated, m_available;
+
+  public:
+    // Emoji();
```
Comment (guerinoni) [inline src/core/Emoji.hpp:18]: Unused

File: src/qml/main.qml [hunk 11]
```diff
@@ -1,23 +1,23 @@
 import QtQuick 2.15
 import QtQuick.Window 2.15
 import QtQuick.Controls 2.15
+//import "Menubar" as Private
```
Comment (guerinoni) [inline src/qml/main.qml:4]: ?

**Review by guerinoni — CHANGES_REQUESTED** [review]
Also is not correctly rebase because it complains about many conflict :^)

**Pull request closed without merging**